src/org/sosy_lab/common/collect/OrderStatisticMap.java

```diff
@@ -0,0 +1,173 @@
+   * Return the rank of the entry with the given key in this map. Returns -1 if the key does not
+   * exist in the map.
+   *
+   * <p>The lowest entry in the set has rank == 0, the largest entry in the set has rank == size -
+   * 1.
+   *
+   * @param pObj the key of the entry to return the rank for
+   * @return the rank of the entry with the given key in the map, or -1 if the key is not in the set
+   * @throws NullPointerException if the given key is <code>null</code>
+   */
+  int rankOf(K pObj);
+
+  @Override
+  OrderStatisticMap<K, V> descendingMap();
+
+  @Override
+  OrderStatisticSet<K> navigableKeySet();
+
+  @Override
+  OrderStatisticSet<K> descendingKeySet();
+
+  @Override
+  OrderStatisticMap<K, V> subMap(K fromKey, boolean fromInclusive, K toKey, boolean toInclusive);
+
+  @Override
+  OrderStatisticMap<K, V> headMap(K toKey, boolean inclusive);
+
+  @Override
+  OrderStatisticMap<K, V> tailMap(K fromKey, boolean inclusive);
+
+  @Override
+  OrderStatisticMap<K, V> subMap(K fromKey, K toKey);
+
+  @Override
+  OrderStatisticMap<K, V> headMap(K toKey);
+
+  @Override
+  OrderStatisticMap<K, V> tailMap(K fromKey);
+
+  /**
+   * Creates a new empty OrderStatisticMap using natural ordering. The returned map guarantees
+   * performance only in O(n) for the operations specific to the OrderStatisticMap interface.
+   */
+  static <K, V> OrderStatisticMap<K, V> create() {
+    return NaiveOrderStatisticMap.createMap();
+  }
+
+  /**
+   * Creates a new empty OrderStatisticMap using the given comparator over its keys. The returned
+   * map guarantees performance only in O(n) for the operations specific to the OrderStatisticMap
+   * interface.
+   */
+  static <K, V> OrderStatisticMap<K, V> create(Comparator<? super K> pComparator) {
+    return NaiveOrderStatisticMap.createMap(pComparator);
+  }
+
+  /**
+   * Creates a new OrderStatisticSet containing the same entries as the given map, using natural
+   * ordering over its keys. The returned map guarantees performance only in O(n) for the operations
+   * specific to the OrderStatisticMap interface.
+   */
+  static <K, V> OrderStatisticMap<K, V> createWithNaturalOrder(Map<? extends K, ? extends V> pMap) {
+    return NaiveOrderStatisticMap.createMapWithNaturalOrder(pMap);
+  }
+
+  /**
+   * Creates a new OrderStatisticMap containing the same entries and using the same order over keys
+   * as the given {@link NavigableMap}. The returned map guarantees performance only in O(n) for the
+   * operations specific to the OrderStatisticMap interface.
+   *
+   * @param pNavigableMap map to use entries and ordering of
+   * @param <K> type of the keys of the given and new map
+   * @param <V> type of the values of the given and new map
+   * @return a new OrderStatisticMap containing the same entries and using the same order over keys
+   *     as the given map
+   */
+  static <K, V> OrderStatisticMap<K, V> createWithSameOrder(
+      NavigableMap<K, ? extends V> pNavigableMap) {
```

Comment:
This parameter could accept a `SortedMap`, right?
(same for `OrderStatisticSet`)